### PR TITLE
One SkinnedMesh Dispatch Per Mesh Part 2: Morph targets

### PR DESCRIPTION
**Pull request by amzn-tommy**
This builds upon the one dispatch per mesh change for skinned meshes to add back support for morph targets. It also fixes a long standing bug where morph targets would not work if they covered more than one mesh. https://github.com/o3de/o3de/issues/3037

Previously, an intermediate buffer the size of the entire lod for the model was created to accumulate deltas from the morph target pass to be applied in the skinned mesh pass. Now, the multiple smaller intermediate buffers are created, and only for meshes that might actually be modified by a morph target, which can represent a significant memory savings for large models that only use morph targets for facial animation (which is pretty common).

Previously, one MorphTargetMetaAsset::MorphTarget was created per 'source mesh' in the builder. But each source mesh might be split into multiple 'product meshes' if there are multiple materials on the mesh. Now, there is one MorphTargetMetaAsset::MorphTarget per 'product mesh', where a product mesh corresponds to an Atom sub-mesh. This still corresponds to a single controllable morph target animation in the animation editor, it is just split into multiple MorphTargetStandard::DeformData's (in emfx parlance) which corresponds to multiple MorphTargetDispatchItem's (in Atom parlance).

The morph target exporter is now also accounting for the vertex re-mapping that is done in the Atom model builder.

I also re-named MorphTargetMetaData to MorphTargetComputeMetaData to differentiate the metadata needed for the compute shader from the MorphTragetMetaAsset::MorphTarget, which are related but not the same.

I tested with a new simplified asset I created to debug the issue with morph targets spread across multiple sub-meshes, as well as with some customer models.

ASV tests pass, with the exception of a known spotlight banding screenshot failure
ASVResults

### CODE REVIEW

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp [hunk 1]
```diff
@@ -552,63 +567,72 @@ namespace AZ
             return true;
         }
 
-        static bool AllocateMorphTargetsForLod(const SkinnedMeshInputLod& lod, size_t vertexCount, AZStd::intrusive_ptr<SkinnedMeshInstance> instance, AZStd::vector<AZStd::intrusive_ptr<SkinnedMeshOutputStreamAllocation>>& lodAllocations)
+        static bool AllocateMorphTargetsForLod(const SkinnedMeshInputLod& lod, AZStd::intrusive_ptr<SkinnedMeshInstance> instance, AZStd::vector<AZStd::intrusive_ptr<SkinnedMeshOutputStreamAllocation>>& lodAllocations)
```
Comment (amzn-tommy) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp:570]: This is actually relatively unchanged compared to the diff, it's just creating multiple intermediate accumulation buffers per lod (one for each mesh that has one or more morph targets)

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshRenderProxy.cpp [hunk 2]
```diff
@@ -53,6 +53,41 @@ namespace AZ
             return true;
         }
 
+        static AZStd::vector<float> CalculateMorphTargetIntegerEncodingsForLod(uint32_t modelLodIndex, Data::Instance<SkinnedMeshInputBuffers> skinnedMeshInputBuffers)
```
Comment (amzn-tommy) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshRenderProxy.cpp:56]: Here instead of having a single encoding per lod, we now calculate one for each mesh.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h [hunk 3]
```diff
@@ -132,6 +144,19 @@ namespace AZ
             //! Calls RPI::Buffer::WaitForUpload for each buffer in the lod.
             void WaitForUpload();
 
+            struct MorphIndex
+            {
+                uint32_t m_meshIndex = 0;
+                uint32_t m_morphIndex = 0;
```
Comment (amzn-jillich) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h:150]: What is the morph index here? The morph target index? Or something else?

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h [hunk 4]
```diff
@@ -176,11 +201,9 @@ namespace AZ
...
             Data::Instance<RPI::Buffer> m_indexBuffer;
 
-            //! Container with one MorphTargetMetaData per morph target that can potentially be applied to an instance of this skinned mesh
-            AZStd::vector<MorphTargetMetaData> m_morphTargetMetaDatas;
-
-            //! Container with one MorphTargetInputBuffers per morph target that can potentially be applied to an instance of this skinned mesh
-            AZStd::vector<AZStd::intrusive_ptr<MorphTargetInputBuffers>> m_morphTargetInputBuffers;
+            //! Keep track of the order in which morph targets were added, so that the weights coming from the animation
+            //! system will be in sync with the morph target dispatch items
+            AZStd::vector<MorphIndex> m_morphDispatchOrder;
```
Comment (amzn-jillich) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h:206]: Why is that needed? The order should be determined by the asset already, right?

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshRenderProxy.cpp (hunk 2 above)
Comment (amzn-jillich) [reply Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshRenderProxy.cpp:56]: That gives us more precision I assume and is now possible as we separately dispatch the morph commands per mesh?

**Review by amzn-jillich — COMMENTED** [review]
Great changes!

Makes me wonder why we were lucky and how this worked before 😅 .

**Review by amzn-jillich — APPROVED** [review]

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h (hunk 4 above)
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h:206]: I need two things, the order in used in EMFX and also to know all of the morph targets that are applied to a given mesh (both knowing if any morph targets are applied to the mesh, and for calculating the integer encoding.

Iterating like
A: for each mesh
  for each morph target dispatch
is a different order than EMFX uses, which is more like
B: for each morph target (MorphTargetStandard)
  for each mesh (MorphTargetStandard::DeformData)
  
Right now I'm storing the MorphTargetComputeMetaDatas for each mesh. In SkinnedMeshRenderProxy, I iterate in order A, and use the dispatch order to track order B. I do think it would be slightly better if I flipped that, and instead of storing the data per-mesh and then using m_morphDispatchOrder to find the order they were created, I could instead store them in the order they were created, and then per-mesh, then iterate over that to calculate the integer encoding per-mesh and determine if any morphs are used for a given mesh at all.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h (hunk 3 above)
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h:150]: This is 'among all the morph target dispatches that could impact the given mesh, which one is it'. As I mentioned in my other comment, I'm going to try flipping this mapping around.

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshRenderProxy.cpp (hunk 2 above)
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshRenderProxy.cpp:56]: Yes, we can get more precision since we're not lumping all morph targets for the whole lod into one encoding.

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp [hunk 5]
```diff
@@ -397,21 +396,58 @@ namespace AZ
...
-        
-        const AZStd::vector<MorphTargetComputeMetaData>& SkinnedMeshInputLod::GetMorphTargetComputeMetaDatas(uint32_t meshIndex) const
+
+        const AZStd::vector<AZStd::intrusive_ptr<MorphTargetInputBuffers>>& SkinnedMeshInputLod::GetMorphTargetInputBuffers() const
         {
-            return m_meshes[meshIndex].m_morphTargetComputeMetaDatas;
+            return m_morphTargetInputBuffers;
+        }
+
+        void SkinnedMeshInputLod::CalculateMorphTargetIntegerEncodings()
```
Comment (amzn-tommy) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp:412]: Moved from SkinnedMeshRenderProxy.cpp, and updated to operate over all meshes in the morph target order instead of one mesh at a time.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h (hunk 4 above)
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/SkinnedMesh/SkinnedMeshInputBuffers.h]: I made the switch with the latest commit, and I think it works out a bit cleaner this way.

File: Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp [hunk 6]
```diff
@@ -397,21 +396,58 @@ namespace AZ
...
+                    // There are no morph targets for this mesh
+                    ranges[i] = -1.0f;
+                }
+                else
+                {
+                    // Given a conservative maximum value of a delta (minimum if negated), set a value for encoding a float as an integer that maximizes precision
+                    // while still being able to represent the entire range of possible offset values for this instance
+                    // For example, if at most all the deltas accumulated fell between a -1 and 1 range, we'd encode it as an integer by multiplying it by 2,147,483,647.
+                    // If the delta has a larger range, we multiply it by a smaller number, increasing the range of representable values but decreasing the precision
+                    m_meshes[i].m_morphTargetIntegerEncoding = static_cast<float>(std::numeric_limits<int>::max()) / ranges[i];
```
Comment (amzn-jillich) [inline Gems/Atom/Feature/Common/Code/Source/SkinnedMesh/SkinnedMeshInputBuffers.cpp:443]: Nit: We might want to rename this into quantization range/delta or something. Needed to check the comment in order to understand what the purpose of this here is.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/MorphTargets/MorphTargetInputBuffers.h [hunk 7]
```diff
@@ -52,14 +52,14 @@ namespace AZ
...
-        struct MorphTargetMetaData
+        struct MorphTargetComputeMetaData
         {
             float m_minWeight;
             float m_maxWeight;
             float m_minDelta;
             float m_maxDelta;
             uint32_t m_vertexCount;
-            uint32_t m_positionOffset;
+            uint32_t m_meshIndex;
```
Comment (Adi-Amazon) [inline Gems/Atom/Feature/Common/Code/Include/Atom/Feature/MorphTargets/MorphTargetInputBuffers.h:62]: nit: remark about what this index is

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp [hunk 8]
```diff
@@ -150,10 +158,17 @@ namespace AZ::RPI
...
+        // The start index is after any previously added deltas from other morph targets
+        // that are part of the same product mesh
+        metaData.m_startIndex = aznumeric_cast<uint32_t>(packedCompressedMorphTargetVertexData.size());
 
         // Determine the vertex index range for the morph target.
-        const uint32_t numVertices = blendShapeData->GetVertexCount();
-        if (blendShapeData->GetVertexCount() != sourceMesh.m_meshData->GetVertexCount())
+        // Ignore any vertices that are associated with a later product mesh
+        uint32_t endVertex = startVertex + aznumeric_cast<uint32_t>(productMesh.m_positions.size() / 3);
+        const uint32_t numVertices = endVertex - startVertex;
```
Comment (Adi-Amazon) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp:169]: super nit: 
Can revert the order of the math since it's repeating itself:
numVertices = m_positions.size() / 3;
endVertex = startVertex + numVertices;

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp [hunk 9]
```diff
@@ -196,7 +208,15 @@ namespace AZ::RPI
                 RPI::CompressedMorphTargetDelta& currentDelta = compressedDeltas.back();
 
                 // Set the morphed index
-                currentDelta.m_morphedVertexIndex = vertexIndex + vertexOffset;
+                const auto& iter = oldToNewIndicesMap.find(vertexIndex);
```
Comment (Adi-Amazon) [inline Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp:211]: nit - quick question to the code above: are we only morphing positions? we are comparing position epsilon but is it possible that a morph will affect normals in our scheme, hence for example create wrinkles without moving all vertices?

**Review by Adi-Amazon — APPROVED** [review]

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp (hunk 9 above)
Comment (amzn-tommy) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp:211]: We do also morph normals, tangents, and bitangents. So it is possible that someone creates a morph that would only modify normals, and it would get skipped in our pipeline if the position didn't move. I'm not sure if that's ever used in practice though. For wrinkles, they would use a second normal map that blends in via a wrinkle mask rather than try to perturb the vertex normal without moving the vertices.

File: Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp (hunk 8 above)
Comment (amzn-tommy) [reply Gems/Atom/RPI/Code/Source/RPI.Builders/Model/MorphTargetExporter.cpp:169]: Done.

File: Gems/Atom/Feature/Common/Code/Include/Atom/Feature/MorphTargets/MorphTargetInputBuffers.h (hunk 7 above)
Comment (amzn-tommy) [reply Gems/Atom/Feature/Common/Code/Include/Atom/Feature/MorphTargets/MorphTargetInputBuffers.h:62]: Done.
